File: numguess.cs
using System;

namespace numguess
{
	class MainClass
	{
		public static void Main(string[] args)
		{
			int limit, tries, number, guess;
			string name, message;
			Random random = new Random();

			Console.Write("Welcome to NumGuess C# version!\n\nEnter your name: ");
			name = Console.ReadLine();
			if (name.Length == 0) name = "Player";

			Console.Write("\nWelcome " + name + ", enter upper limit: ");

			try
			{
				limit = int.Parse(Console.ReadLine());
				if (limit < 10) limit = 10;
			}
			catch
			{
				limit = 10;
			}

			int max_tries = (int)Math.Floor(Math.Log(limit) / Math.Log(2)) + 1;

			do
			{
				tries = 0;
				number = random.Next(1, limit + 1);

				Console.WriteLine("\nGuess my number between 1 and " + limit + "!\n");

				do
				{
					Console.Write("Guess: ");

					try
					{
						guess = int.Parse(Console.ReadLine());
						if (guess > limit || guess < 1)
						{
							Console.WriteLine("Out of range.");
						}
						else
						{
							tries++;
							if (guess < number) Console.WriteLine("Too low!");
							else if (guess > number) Console.WriteLine("Too high!");
						}
					}
					catch
					{
						guess = 0;
						Console.WriteLine("That's just plain wrong.");
					}

				} while (guess != number);

				Console.WriteLine("\nWell done " + name + ", you guessed my number from " + tries + " " + (tries == 1 ? "try" : "tries") + "!");

				if (tries == 1)
					message = "You're one lucky bastard!";
				else if (tries < max_tries)
					message = "You are the master of this game!";
				else if (tries == max_tries)
					message = "You are a machine!";
				else if (tries <= max_tries * 1.1)
					message = "Very good result!";
				else if (tries <= limit)
					message = "Try harder, you can do better!";
				else
					message = "I find your lack of skill disturbing!";

				Console.WriteLine(message);
				Console.Write("Play again [y/N]? ");

			} while (Console.ReadLine().ToUpper().Equals("Y"));

			Console.WriteLine("\nOkay, bye.");
		}
	}
}

File: numguess.dart
import 'dart:async';
import 'dart:convert';
import 'dart:io';
import 'dart:math';

void main() {
	new NumGuess();
}

class NumGuess {
	String state;
	String name;
	int limit;
	int num;
	int tries;
	int max_tries;
	
	var random;
	
	NumGuess() {
		random = new Random();
		
		stdout.write('Welcome to NumGuess Dart version!\n\nEnter your name: ');
		state = 'name';

		var sub;
		sub = UTF8.decoder.bind(stdin).listen((str) {
			str = str.trim();
			switch(state) {
				case 'name':
					name = str == '' ? 'Player' : str;
					stdout.write('\nWelcome ${name}, enter upper limit: ');
					state = 'limit';
					break;
					
				case 'limit':
					limit = max(10, int.parse(str, radix:10, onError:(str) => 10));
					max_tries = (log(limit) / log(2)).floor() + 1;
					startGame();
					break;
					
				case 'guess':
					if(evaluateGuess(str)) wellDone(); else stdout.write('Guess: ');
					break;
					
				case 'play':
					if(str.toUpperCase() == 'Y') startGame(); else {
						sub.cancel();
						stdout.writeln('\nOkay, bye.');
						exit;
					}
					break;
			}
		});
	}
	
	void startGame() {
		tries = 0;
		num = random.nextInt(limit) + 1;
		stdout.write('\nGuess my number between 1 and ${limit}!\n\nGuess: ');
		state = 'guess';
	}
	
	bool evaluateGuess(String str) {
		int guess = int.parse(str, radix:10, onError:(str) => null);
		if(guess == null) stdout.writeln('That\'s just plain wrong.');
		else if(guess < 1 || guess > limit) stdout.writeln('Out of range.');
		else {
			tries++;
			if(guess < num) stdout.writeln('Too low!');
			else if(guess > num) stdout.writeln('Too high!');
			else return true;
		}
		return false;
	}
	
	void wellDone() {
		String tries_word = tries == 1 ? 'try' : 'tries';
		stdout.writeln('\nWell done ${name}, you guessed my number from ${tries} ${tries_word}!');
		
		if(tries == 1) stdout.writeln('You\'re one lucky bastard!');
		else if(tries < max_tries) stdout.writeln('You are the master of this game!');
		else if(tries == max_tries) stdout.writeln('You are a machine!');
		else if(tries <= max_tries * 1.1) stdout.writeln('Very good result!');
		else if(tries > limit) stdout.writeln('I find your lack of skill disturbing!');
		else stdout.writeln('Try harder, you can do better!');
		
		stdout.write('Play again [Y/N]? ');
		state = 'play';
	}
}
